samples: basic: minimal: Add a config for no SW ISR table

bjarki-andreasen opened the pull request:
Extended basic/minimal sample and added fixes which allow for and ensure we can build supported targets with `CONFIG_MULTITHREADING=n` and `CONFIG_GEN_SW_ISR_TABLE=n` to achieve the absolute leanest application possible capable of running hello world:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       10976 B      1428 KB      0.75%
             RAM:        3456 B       188 KB      1.80%
        IDT_LIST:          0 GB        32 KB      0.00%
```
The sample discovered a bug in `isr_tables.c`, which was introduced as a fix for MDWT, so we need some proper fix for the issue, this PR adds a conditional which applies the mdwt fix only for the arcmdwt toolchain.

bjarki-andreasen-nordic committed:
drivers: serial: nrfx_uarte: Default DIRECT_ISR if MULTITHREADING=n

Default to using DIRECT_ISR if nothreading. This allows the SW ISR
table to be excluded if nothreading.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
Please take a look @evgeniy-paltsev as this PR is reverting a fix for MWDT, we need a proper fix I imagine which works for both GCC and MWDT :)

masz-nordic commented:
Could be a custom config for `samples/basic/minimal` instead of introducing a new sample?

kartben reviewed (changes requested):
Samples need a README
Also not sure why we need a sample for this in the first place?

nashif commented:
> The sample discovered a bug in `isr_tables.c`, which was introduced as a fix for MDWT, so we need some proper fix for the issue, this PR simply reverts the previous fix, lets see what happens :)

hmm, is this our modus operandi now? If this is an issue/bug, open one and get the person who introduced this a chance to comment, or at least add them to the view here asking for input etc. @ruuddw @evgeniy-paltsev can you please takea look?

bjarki-andreasen-nordic committed:
samples: basic: minimal: add no-sw-isr-table option

Add option to build without software isr table for platforms which
support it (ARM and RISCV). The software ISR table takes up
kilobytes of RAM and introduces latency, and it is not needed if
CONFIG_MULTITHREADING=n.

The initially supported nrf54l15 and nrf54lm20 boards additionally
need CONFIG_CLOCK_CONTROL=n for the minimal sample so board configs
have been added for them. Note that the entry in sample.yaml uses
EXTRA_CONF_FILE to not overwrite the board specific CONF_FILE.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
> > The sample discovered a bug in `isr_tables.c`, which was introduced as a fix for MDWT, so we need some proper fix for the issue, this PR simply reverts the previous fix, lets see what happens :)
> 
> hmm, is this our modus operandi now? If this is an issue/bug, open one and get the person who introduced this a chance to comment, or at least add them to the view here asking for input etc. @ruuddw @evgeniy-paltsev can you please takea look?

@anasnashif maybe you should look to the comments before throwing around accusations https://github.com/zephyrproject-rtos/zephyr/pull/94132#issuecomment-3155982458

bjarki-andreasen commented on samples/boards/nordic/nothreading/sample.yaml:
Yep, @masz-nordic notes something similar, I have extended the basic/minimal sample :)

bjarki-andreasen commented:
> Samples need a README Also not sure why we need a sample for this in the first place?

We don't need a new sample it seems, as noted by a few reviewers, extended basic/minimal instead which seems perfect :)

carlescufi commented:
Apologies, clicked on Close by mistake

bjarki-andreasen commented:
> @bjarki-andreasen does this fix #86311 ?

No, CONFIG_GEN_ISR_TABLES is needed by at least all ARM cortex_m platforms, otherwise the IRQ vector table is not generated. This PR only allows disabling CONFIG_GEN_SW_ISR_TABLE, (keeping CONFIG_GEN_IRQ_VECTOR_TABLE).

bjarki-andreasen commented:
> > > @bjarki-andreasen does this fix #86311 ?
> > 
> > 
> > No, CONFIG_GEN_ISR_TABLES is needed by at least all ARM cortex_m platforms, otherwise the IRQ vector table is not generated. This PR only allows disabling CONFIG_GEN_SW_ISR_TABLE, (keeping CONFIG_GEN_IRQ_VECTOR_TABLE).
> 
> Hmm... I got it working on other cortex_m boards but not on Nordic. So this is kinda surprising to me, but maybe there is something else differing between vendors I am missing. But if this is true that cortex_m implies GEN_ISR_TABLES maybe that should be encoded into Kconfig. However it has been ages since I looked at this so I don't remember details, just thought I would ask to learn :)

I replied on the issue itself, if no driver or soc code causes any interrupt, the sample can run :) That's likely what you are observing

carlescufi commented:
> > It's not a new sample, it's just one more configuration of the minimal sample,
> 
> It was a new Nordic board sample when I commented...

Right, that's what I suspected, and why I wrote "@kartben I assume @bjarki-andreasen has reworked this after you commented, but just in case:"

nashif reviewed (changes requested):
requesting review from @evgeniy-paltsev and @ruuddw

bjarki-andreasen commented:
> > @anasnashif maybe you should look to the comments before throwing around accusations [#94132 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/94132#issuecomment-3155982458)
> 
> still not added as a reviewer, stakeholders should be added as a reviewer and this PR should be imo DNM until you get the input.

Actually, given the PR is green, as in, no regressions have been found in CI, there is nothing blocking it from moving forward, and a DNM would in fact be unfounded. I'm actively bringing up a POTENTIAL regression which only MAY exist, to the person who would know how to make sure no regression exists. Everything is entirely up to code, going even beyond what is required of any PR.

bjarki-andreasen commented:
ping @evgeniy-paltsev @kartben

bjarki-andreasen commented on drivers/timer/nrf_grtc_timer.c:
```diff
@@ -459,12 +459,35 @@ uint32_t sys_clock_elapsed(void)
 	return (uint32_t)(counter_sub(counter(), last_count) / CYC_PER_TICK);
 }
 
+#if CONFIG_NRF_GRTC_TIMER_ISR_DIRECT
```
I mean, a search for `#if CONFIG_` gives ~8000 matches, `#ifdef CONFIG_` > 20000, don't really think there is much of a convention. kconfig is specifically built to work with both, so really it does not matter.

ruuddw commented:
From the (5 year old) commit that's reverted I see the problem the MetaWare linker signals probably should have been signalled by the gnu linker as well.

> On the final linkage stage we pass two files which content
> section with sw_isr_table / irq_vector_table. They are
>  * libarch__common.a (with an outdated tables from the first
>    linkage stage)
>  * isr_tables.c.obj (with an actual tables)
> 
> The sections where tables are located are marked with
> ".gnu.linkonce" prefix. That means:
> <<<As a GNU extension, if the name begins with .gnu.linkonce,
> we only link a single copy of the section.>>>
> 
> However the "libarch__common.a" is passed to linker with
> "--whole-archive" option which means <<<include every object
> file in the archive in the link, rather than searching the archive
> for the required object files>>>
> 
> That combination confuses MWDT linker and breaks linkage with
> MWDT toolchain.

The MetaWare linker is based on the llvm lld linker. Is MetaWare the the only non-gnu toolchain that generates a duplicate symbol error?
```
lldac: error: duplicate symbol: _irq_vector_table
>>> defined at isr_tables.c:10 (C:/workspaces/zephyrproject/zephyr/build/zephyr\isr_tables.c:10)
>>>            zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj:(_irq_vector_table)
>>> defined at isr_tables.c:81 (C:/workspaces/zephyrproject/zephyr/arch/common\isr_tables.c:81)
>>>            isr_tables.c.obj:(.gnu.linkonce.irq_vector_table._irq_vector_table+0x0) in archive zephyr/arch/common/libarch__common.a

```

ruuddw reviewed (changes requested):
Breaks Metaware builds.

bjarki-andreasen commented:
> > > Breaks Metaware builds.
> > 
> > 
> > Right, and the original commit breaks ARM builds, so we need to find a solution which works for both, the obvious fix (in pseudo code) is to add:
> > ```
> > if(NOT Metaware)
> >   <zephyr lib>
> > else()
> >   <non zephyr lib>
> > endif()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > to the cmake file, not the prettiest, but should work.
> 
> I believe the actual reason for the failure is that non-GNU linkers don't really support the special meaning of .gnu.linkonce sections. I wonder if we could just get rid of that and mark the non-generated table as weak instead. Would it work?

Maybe, I'm trying to introduce as few changes as possible, the section magic we do is fragile enough as is, if we know what works for each linker, we can keep it simple no?

bjarki-andreasen commented:
@ruuddw Reverted the logic, the toolchain variant used which fails is "zephyr"

bjarki-andreasen commented on arch/common/CMakeLists.txt:
```diff
@@ -95,17 +95,19 @@ if (DEFINED CONFIG_ARCH_SUPPORTS_ROM_OFFSET)
-  target_link_libraries(isr_tables zephyr_interface)
-  zephyr_library_link_libraries(isr_tables)
+if(CONFIG_GEN_ISR_TABLES)
+  if(ZEPHYR_TOOLCHAIN_VARIANT STREQUAL "zephyr")
```
I have no idea if it has anything to do with llvm or gcc... What has been observed is:

* arm-zephyr-eabi fails to generate the vector table if the isr_tables.c is not a zephyr library
* arcmwdt fails to link if isr_tables.c is a zephyr library.
* The only toolchain observed to fail if irq_tables.c is a zephyr library is arcmwdt

That's why my original logic checked for arcmwdt and kept the patch for arcmwdt in that case, since that original PR broke arm-zephyr-eabi unknowingly (no-one has built only the irq vector table with arm-zephyr-eabi up until now it seems)

bjarki-andreasen commented:
I reverted the logic in CMakeLists.txt to
```
if(CONFIG_GEN_ISR_TABLES)
  if(ZEPHYR_TOOLCHAIN_VARIANT STREQUAL "arcmwdt")
    # isr_tables must be a normal CMake library and not a zephyr_library because
    # it should not be --whole-archive'd for arcmwdt
    add_library(isr_tables isr_tables.c)
    add_dependencies(isr_tables zephyr_generated_headers)
    target_link_libraries(isr_tables zephyr_interface)
    zephyr_library_link_libraries(isr_tables)
  else()
    # isr_tables must be a zephyr_library as it has to be --whole-archive'd to
    # keep the _iheader, in section .irq_info, included in the build.
    zephyr_library_sources(isr_tables.c)
  endif()
endif()
```
This matches the observations made and solves the issue for ARM with GCC, while not affecting MWDT. Unless this fix breaks any other platforms, please approve the fix in this PR or propose an alternate solution, ARM with GCC is broken without this fix.

bjarki-andreasen commented:
> I'd be ok with this as it works for MetaWare. But there may be others now or in the near future which have the same problem.
> The recommendation from @nashif was to check for gcc.
> Would there be other ways of avoiding symbol `_iheader` to get optimized out?

Actually, the data which is stored in that struct seems entirely dependent on configs, so we might just infer these values from .config rather than creating a struct to store it in in the first place :)

bjarki-andreasen commented:
@RobinKastberg any progress on testing?

RobinKastberg reviewed (changes requested):
Yes.. Seems like IAR needs the same workaround.
However I don't like this fix in general, since it seems to introduce more complications than it solves. 

Since I don't have a proper solution I don't want to block this if it this fix is helpful to someone (as long as the IAR workaround gets added)

carlescufi commented:
@RobinKastberg PTAL

nashif commented:
> Not a surprise that it breaks on IAR, as others and I mentioned earlier, it seems gcc is the exception, not Metaware, IAR and probably other toolchains.

@bjarki-andreasen  did you take this comment into account?

carlescufi reviewed (changes requested):
Actually, I just tried cross-compiling with clang and I get similar errors, but in fact I do get those errors even on other samples with this PR! @tejlmand has been informed

```
export ZEPHYR_TOOLCHAIN_VARIANT=llvm
LLVM_TOOLCHAIN_PATH=/Users/carles/bin/LLVM-ET-Arm-19.1.5-Darwin-universal
west build samples/hello_world -- -DCONFIG_LLVM_USE_LLD=y -DCONFIG_COMPILER_RT_RTLIB=y
```
yields:
```
ld.lld: error: duplicate symbol: _irq_vector_table
>>> defined at isr_tables.c:10 (zephyr/isr_tables.c:10)
>>>            zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj:(_irq_vector_table)
>>> defined at isr_tables.c:81 (/Users/carles/src/zephyr/zephyr/arch/common/isr_tables.c:81)
>>>            isr_tables.c.obj:(.gnu.linkonce.irq_vector_table+0x0) in archive zephyr/arch/common/libarch__common.a

ld.lld: error: duplicate symbol: _sw_isr_table
>>> defined at isr_tables.c:283 (zephyr/isr_tables.c:283)
>>>            zephyr/CMakeFiles/zephyr_final.dir/isr_tables.c.obj:(_sw_isr_table)
>>> defined at isr_tables.c:94 (/Users/carles/src/zephyr/zephyr/arch/common/isr_tables.c:94)
>>>            isr_tables.c.obj:(.gnu.linkonce.sw_isr_table+0x0) in archive zephyr/arch/common/libarch__common.a
clang: error: ld.lld command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /opt/homebrew/bin/cmake --build /Users/carles/src/zephyr/zephyr/build
```

If I revert the commit `arch: common: fix missing _iheader from isr_tables.c for gcc`, then I get the same with both GCC and clang:

```
west build samples/basic/minimal -- -DEXTRA_CONF_FILE='common.conf;no-mt.conf;no-sw-isr-table.conf;arm.conf'
```
yields
```
cd /Users/carles/src/zephyr/zephyr/build/zephyr && /Users/carles/src/zephyr/.venv/bin/python3.12 /Users/carles/src/zephyr/zephyr/scripts/build/gen_isr_tables.py --output-source isr_tables.c --linker-output-files isr_tables_vt.ld isr_tables_swi.ld --kernel /Users/carles/src/zephyr/zephyr/build/zephyr/zephyr_pre0.elf --intlist-section .intList --intlist-section intList --vector-table
Traceback (most recent call last):
  File "/Users/carles/src/zephyr/zephyr/scripts/build/gen_isr_tables.py", line 342, in <module>
    main()
  File "/Users/carles/src/zephyr/zephyr/scripts/build/gen_isr_tables.py", line 322, in main
    parser = parser_module.gen_isr_parser(intlist_data, config, log)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carles/src/zephyr/zephyr/scripts/build/gen_isr_tables_parser_carrays.py", line 41, in __init__
    intlist = self.__read_intlist(intlist_data)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/carles/src/zephyr/zephyr/scripts/build/gen_isr_tables_parser_carrays.py", line 88, in __read_intlist
    struct.iter_unpack(intlist_entry_fmt, intdata)]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
struct.error: iterative unpacking requires a buffer of a multiple of 16 bytes
ninja: build stopped: subcommand failed.
```

bjarki-andreasen commented:
> > Not a surprise that it breaks on IAR, as others and I mentioned earlier, it seems gcc is the exception, not Metaware, IAR and probably other toolchains.
> 
> @bjarki-andreasen did you take this comment into account?

I have taken every comment into account yes, have you seen this comment?

> Since I don't have a proper solution I don't want to block this if it this fix is helpful to someone (as long as the IAR workaround gets added

How about this one https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2280938864

which you never came back to, if there is a better solution, I am, and have always had, open ears

nashif commented:
> > > Not a surprise that it breaks on IAR, as others and I mentioned earlier, it seems gcc is the exception, not Metaware, IAR and probably other toolchains.
> > 
> > 
> > @bjarki-andreasen did you take this comment into account?
> 
> I have taken every comment into account yes, have you seen this comment?
> 
> > Since I don't have a proper solution I don't want to block this if it this fix is helpful to someone (as long as the IAR workaround gets added
> 
> How about this one [#94132 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2280938864)
> 
> which you never came back to, if there is a better solution, I am, and have always had, open ears

you do not need to get testy. 
This comment https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2279987877 and https://github.com/zephyrproject-rtos/zephyr/pull/94132#issuecomment-3361896948 and finally the findings by @carlescufi all point to an issue that this PR does not fully address. If someone is ok with one solution to get over part of the problem does not mean we should not address the main issue.

bjarki-andreasen commented:
> > > > Not a surprise that it breaks on IAR, as others and I mentioned earlier, it seems gcc is the exception, not Metaware, IAR and probably other toolchains.
> > > 
> > > 
> > > @bjarki-andreasen did you take this comment into account?
> > 
> > 
> > I have taken every comment into account yes, have you seen this comment?
> > > Since I don't have a proper solution I don't want to block this if it this fix is helpful to someone (as long as the IAR workaround gets added
> > 
> > 
> > How about this one [#94132 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2280938864)
> > which you never came back to, if there is a better solution, I am, and have always had, open ears
> 
> you do not need to get testy. This comment [#94132 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2279987877) and [#94132 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/94132#issuecomment-3361896948) and finally the findings by @carlescufi all point to an issue that this PR does not fully address. If someone is ok with one solution to get over part of the problem does not mean we should not address the main issue.

At which point was "we should not address the main issue" suggested by anyone? The findings by @carlescufi where made 3 days ago, after 3 weeks of silence on this PR. At the very least I (and likely anyone else following this PR) is not pushing for this PR to get merged with that or any issue present.

I'm annoyed at frequent, unhelpful, and what I perceive to be snarky, comments from you, that's why I'm being rather dismissive of your comments at this point.

nashif commented:
> At which point was "we should not address the main issue" suggested by anyone?

i was referring to this:

> I have taken every comment into account yes, have you seen this comment?

>> Since I don't have a proper solution I don't want to block this if it this fix is helpful to someone (as long as the IAR workaround gets added

Yes, I have seen this comment.

Reason for my question ("did you take this comment into account?") was the fact you pushed another revision addressing that which did not take https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2279987877 and https://github.com/zephyrproject-rtos/zephyr/pull/94132#issuecomment-3361896948 into consideration.


> At the very least I (and likely anyone else following this PR) is not pushing for this PR to get merged with that or any issue present.

Okay, I, and others, are responding to pings both here and on discord's #pr-help, I do not see how my question above becomes so controversial or "snarky". 

> I'm annoyed at frequent, unhelpful, and what I perceive to be snarky, comments from you, that's why I'm being rather dismissive of your comments at this point.

I apologize if my comments above came as "snarky" or "unhelpful".

tejlmand committed:
cmake: define linker argument for undefined symbol

Specifying undefined symbol on linker invocation requests the linker to
search and include the symbol.
This ensures the symbol will be present in final output.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand commented:
Pushed a new commit which changes the principle into the toolchaing (linker) property scheme which is the prefered way for supporting multiple toolchains.

@evgeniy-paltsev please take a look from arcmwdt point.
@RobinKastberg I'm specifying `--keep` for IAR, afaiu that's the most correct flag to use for ilinkarm, but please provide alternative if you don't believe this to be correct.

tejlmand commented on samples/basic/minimal/sample.yaml:
```diff
@@ -105,3 +105,15 @@ tests:
     integration_platforms:
       - qemu_riscv32
+  sample.minimal.no-mt-no-sw-isr-table.arm:
+    extra_args: EXTRA_CONF_FILE='common.conf;no-mt.conf;no-sw-isr-table.conf;arm.conf'
```
line related to comment: https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2410028659

carlescufi commented on samples/basic/minimal/no-sw-isr-table.conf:
```diff
@@ -0,0 +1,3 @@
+# No software ISR table, only direct ISR support in the kernel
+
+CONFIG_GEN_SW_ISR_TABLE=n
```
But this is the pattern used in this sample, to ensure that we can always refer to `.conf` files even if Kconfig options change in the future. This is critical for this particular sample since it's supposed to always show the absolute minimal config at all times.

carlescufi reviewed (dismissed):
This works fine now with both GCC and Clang

tejlmand reviewed (dismissed):
Approved, as discussion here https://github.com/zephyrproject-rtos/zephyr/pull/94132#discussion_r2410028659 is non-blocking.

ruuddw reviewed (dismissed):
The "-u" linker property works for Metaware, I think this is a better solution.

nashif reviewed (dismissed):
@RobinKastberg  please take another look

RobinKastberg reviewed (changes requested):
Sorry, I haven't had time to take a proper look, but testing:

I get the following:
``` /workdir/5216/arm/bin/ilinkarm --no-wrap-diagnostics --entry_list_in_address_order --log modules,libraries,initialization,redirects,sections  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.o  --config=/workdir/zephyr/build/zephyr/linker_zephyr_pre0.cmd  --map=/workdir/zephyr/build/zephyr/zephyr_pre0.map  --log_file=/workdir/zephyr/build/zephyr/zephyr_pre0.map.log  --whole_archive  app/libapp.a  --whole_archive  zephyr/libzephyr.a  --whole_archive  zephyr/arch/common/libarch__common.a  --whole_archive  zephyr/arch/arch/arm/core/libarch__arm__core.a  --whole_archive  zephyr/arch/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a  --whole_archive  zephyr/lib/libc/minimal/liblib__libc__minimal.a  --whole_archive  zephyr/lib/libc/common/liblib__libc__common.a  --whole_archive  zephyr/soc/soc/ti_lm3s6965/libsoc__ti__lm3s6965.a  --whole_archive  zephyr/drivers/console/libdrivers__console.a  --whole_archive  zephyr/drivers/serial/libdrivers__serial.a  --whole_archive  zephyr/drivers/timer/libdrivers__timer.a  zephyr/kernel/libkernel.a  --entry=__start  --diag_suppress=Lt056  -f /workdir/zephyr/build/zephyr/linker_zephyr_pre0.cmd.xcl  zephyr/arch/common/libisr_tables.a  --keep_sw_isr_table  --keep_irq_vector_table zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.o -o zephyr/zephyr_pre0.elf

   IAR ELF Linker V9.70.1.5216/LNX [NIGHTLY BUILD] for ARM
   Copyright 2007-2025 IAR Systems AB.
Command line error: Unexpected command line arguments:
            --keep_sw_isr_table
            --keep_irq_vector_table
```

Changing `--keep` to `--keep=`  manually in command line gives

makes linking pass (I have not verified further yet)

RobinKastberg reviewed (approved):
LGTM
